src/Cassandra/TcpSocket.cs

```diff
@@ -389,39 +392,47 @@ protected virtual void OnClosing()
             _receiveBuffer = null;
         }
 
+        private void OnWriteFlushed()
+        {
+            var callback = Interlocked.Exchange(ref _writeFlushCallback, null);
+            if (callback != null)
+            {
+                callback();
+            }
+        }
+
         /// <summary>
         /// Sends data asynchronously
         /// </summary>
-        public virtual void Write(Stream stream)
+        public void Write(byte[] buffer, int length, Action onBufferFlush)
         {
-            //This can result in OOM
-            //A neat improvement would be to write this sync in small buffers when buffer.length > X
-            var buffer = Utils.ReadAllBytes(stream, 0);
+            Interlocked.Exchange(ref _writeFlushCallback, onBufferFlush);
             if (_isClosing)
             {
                 OnError(new SocketException((int)SocketError.Shutdown));
+                OnWriteFlushed();
                 return;
             }
             if (_sendSocketEvent != null)
             {
-                _sendSocketEvent.SetBuffer(buffer, 0, buffer.Length);
-                bool willRaiseEvent = false;
+                _sendSocketEvent.SetBuffer(buffer, 0, length);
```

Comment:
Remember that large buffers are pooled, meaning that few large buffers are created and then reused for writing across all connections.
`GetBuffer()` writes small chunks into a "final" large buffer that is used to send. It is indeed the same as writing to 1 buffer (the `SocketAsyncEventArgs` buffer) per connection, while avoiding the multiple system calls in case the amount of messages is large.

I think I found a solution that would avoid the block copying from stream to final buffer,  using [SocketAsyncEventArgs.BufferList](https://msdn.microsoft.com/en-us/library/system.net.sockets.socketasynceventargs.bufferlist(v=vs.110).aspx). I will try that when using the `SocketAsyncEventArgs` interface.